### PR TITLE
support MultiTarget when require_local_access is True

### PR DESCRIPTION
**Pull request by FHoffmannCode**

### CODE REVIEW

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py [hunk 1]
```diff
@@ -27,25 +33,46 @@ def __init__(self, task_run):
...
-                    os.path.basename(p_val.path),
-                    config=p_val.config,
+                local_target = self._sync_local_target(p_val)
+
+                self._sync_input_or_output(p_def, p_val, local_target)
+
+            elif (
+                isinstance(p_val, MultiTarget)
+                and p_val.config.require_local_access
+                and not p_val.fs.local
```
Comment (evgenyshulman) [inline modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:43]: we probably should check every target in multitarget for fs.local

**Review by evgenyshulman — COMMENTED** [review]
we need some simple logic, that if at least one target was converted into local -> we create a new target, otherwise, we can keep an old one.

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py [hunk 2]
```diff
@@ -27,25 +33,46 @@ def __init__(self, task_run):
...
-                    # Use DbndLocalFileMetadataRegistry to sync inputs only when necessary
-                    dbnd_meta_cache = DbndLocalFileMetadataRegistry.get_or_create(
-                        local_target
-                    )
 
-                    # When syncing remote to local -> We can't use MD5, so use TTL instead
-                    if dbnd_meta_cache.expired or not local_target.exists():
-                        # If TTL is invalid, or local file doesn't exist -> Sync to local
-                        self.inputs_to_sync.append((p_def, p_val, local_target))
+    def _sync_local_target(self, target_):
```
[the remark this answers is not in the document]
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py]: sure, fixd

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py (hunk 1 above)
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py]: yes, my bad, I've refactored this logic

**Review by evgenyshulman — CHANGES_REQUESTED** [review]

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py [hunk 3]
```diff
@@ -27,25 +33,39 @@ def __init__(self, task_run):
...
-                    )
+                local_target = self._local_cache_target(p_val)
+
+                self._sync_input_or_output(p_def, p_val, local_target)
+
+            elif isinstance(p_val, MultiTarget):
+                for target_ in p_val.targets:
+                    if target_.config.require_local_access and not target_.fs.local:
+                        local_target = self._local_cache_target(target_)
+                        self._sync_input_or_output(p_def, target_, local_target)
```
[the remark this answers is not in the document]
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py]: you commented on outdated version of the code, in current version this line is gone

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py [hunk 4]
```diff
@@ -27,72 +33,137 @@ def __init__(self, task_run):
...
-                    # Use DbndLocalFileMetadataRegistry to sync inputs only when necessary
-                    dbnd_meta_cache = DbndLocalFileMetadataRegistry.get_or_create(
-                        local_target
+                self._sync_input_or_output(p_def, p_val, local_target)
+
+            elif isinstance(p_val, MultiTarget):
+                remote_targets_to_be_cached = []
+                local_targets = []
+                for target_ in p_val.targets:
+                    if target_.config.require_local_access and not target_.fs.local:
```
[the remark this answers is not in the document]
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:44]: are you sure? from what I saw in code `MultiTarget` gets its config from first subtarget on its `targets` attribute so if we're 'unlucky' and this first subtarget does not have `config.require_local_access` set to `True` then we wont sync any subtargets, even the remote ones

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py [hunk 5]
```diff
@@ -27,25 +33,39 @@ def __init__(self, task_run):
...
+                local_target = self._local_cache_target(p_val)
+
+                self._sync_input_or_output(p_def, p_val, local_target)
+
+            elif isinstance(p_val, MultiTarget):
+                for target_ in p_val.targets:
+                    if target_.config.require_local_access and not target_.fs.local:
+                        local_target = self._local_cache_target(target_)
+                        self._sync_input_or_output(p_def, target_, local_target)
+                        local_target.append(local_target)
```
[the remark this answers is not in the document]
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py]: no, its just an outdated piece of code

[the remark this answers is not in the document]
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py]: yes but its an old and outdated code

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py [hunk 6]
```diff
@@ -27,72 +33,137 @@ def __init__(self, task_run):
...
+    def _download_remote_target_to_local_path(remote_target, local_target):
+        # type: (Type[FileTarget], Type[FileTarget]) -> None
+        local_target.mkdir_parent()
+        remote_target.download(
+            local_target.path, overwrite=local_target.config.overwrite_target,
+        )
+
+    def _copy_local_multitarget_to_remote(self, remote_multitarget, local_multitarget):
+        # type: (MultiTarget, MultiTarget) -> None
+        for remote_target, local_target in zip(
```
[the remark this answers is not in the document]
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:159]: I'm not sure what you mean. This function is used here:

```
    def sync_post_execute(self):
        for p_def, remote_target, local_target in self.inputs_to_sync:
            setattr(self.task, p_def.name, remote_target)

        for p_def, remote_target, local_target in self.outputs_to_sync:
            try:
                logger.info(
                    "Uploading  %s %s from %s", p_def, remote_target, local_target
                )
                if isinstance(remote_target, MultiTarget):
                    self._copy_local_multitarget_to_remote(remote_target, local_target)
                else:
                    self._copy_local_target_to_remote(remote_target, local_target)
```

do you want me to move uploading/downloading logic back to `sync_post_execute` function?

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (evgenyshulman) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:44]: you are correct, I was sure it relates to ParameterDefinition, so let's use your way now, 
and in extra task, we move this property into ParameterDefinition.

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (evgenyshulman) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:159]: I would suggest to move logic that is in charge of setting the attribute and downloading -> into pre_execute
the logic that is in charge of reverting -> in post_execute

[the remark this answers is not in the document]
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:159]: I'll do it if you insist however IMO `pre_execute is already pretty long and has a lot of nested code which makes it difficult to read. If I move all download/upload logic back there, the readability will worsen.

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py [hunk 7]
```diff
@@ -27,72 +33,137 @@ def __init__(self, task_run):
...
+        )
+
+    def _sync_input_or_output(self, param_definition, old_target, new_target):
+        # type: (ParameterDefinition, Type[DataTarget], Type[DataTarget]) -> None
+        if param_definition.kind == _ParameterKind.task_output:
+            # Output should be substituted for local path and synced post execution
+            self.outputs_to_sync.append((param_definition, old_target, new_target))
+
+        else:
+            # Use DbndLocalFileMetadataRegistry to sync inputs only when necessary
```
[the remark this answers is not in the document]
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:72]: ok, done

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py [hunk 8]
```diff
@@ -27,72 +33,137 @@ def __init__(self, task_run):
...
+                for target_ in p_val.targets:
+                    if target_.config.require_local_access and not target_.fs.local:
+                        remote_targets_to_be_cached.append(target_)
+                        local_targets.append(self._local_cache_target(target_))
+
+                if local_targets:
+                    # A new MultiTarget with remote targets to be cached is needed
+                    # so that remote targets can be downloaded to local path later.
+                    remote_multitarget = MultiTarget(remote_targets_to_be_cached)
+                    local_multitarget = MultiTarget(local_targets)
```
[the remark this answers is not in the document]
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:52]: ok fixd, I hope I got this right this time

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py [hunk 9]
```diff
@@ -27,72 +33,140 @@ def __init__(self, task_run):
...
-                except Exception as ex:
+                    if isinstance(remote_target, MultiTarget):
+                        for remote_subtarget, local_subtarget in zip(
+                            remote_target.targets, local_target.targets
+                        ):
+                            if (
+                                remote_subtarget.config.require_local_access
+                                and not remote_subtarget.fs.local
+                            ):
+                                # Use DbndLocalFileMetadataRegistry to sync inputs only when necessary
```
Comment (evgenyshulman) [inline modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:85]: all this code is just :  sync_remote_to_local(remote_target). (That can be used by Multi and regular). -- returns local target

**Comment by evgenyshulman** [comment]
@FHoffmannCode  we can probably test all these code , by creating task with Targets, that have FileSystem that  inherts from local, but marked as not local (pseudo -not local) --> so the system will download/upload files from/into it

File: modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py (hunk 9 above)
Comment (FHoffmannCode) [reply modules/dbnd/src/dbnd/_core/task_run/task_run_sync_local.py:85]: true, fixd

**Comment by evgenyshulman** [comment]
LGTM. please merge.

**Review by evgenyshulman — APPROVED** [review]
we still at least one test. we can do it in separate PR.

**Comment by FHoffmannCode** [comment]
@evgenyshulman I'm currently writing tests, should be ready today

**Pull request merged**